### PR TITLE
Add a msg for create PeriodicVestingAccount

### PR DESCRIPTION
**Pull request by zmanian**
Pull Request of the periodic vesting msg

## Description

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: proto/cosmos/vesting/v1beta1/tx.proto [hunk 1]
```diff
@@ -28,4 +30,15 @@ message MsgCreateVestingAccount {
...
+
+// MsgCreateVestingAccount defines a message that enables creating a vesting
+// account.
+message MsgCreatePeriodicVestingAccount {
+  option (gogoproto.equal) = false;
+
+  string   from_address                    = 1 [(gogoproto.moretags) = "yaml:\"from_address\""];
+  string   to_address                      = 2 [(gogoproto.moretags) = "yaml:\"to_address\""];
+  repeated Period vesting_periods = 3 [(gogoproto.nullable) = false];
+}
```
Comment (JimLarson) [inline proto/cosmos/vesting/v1beta1/tx.proto]: Add newline at end of file.

File: proto/cosmos/vesting/v1beta1/tx.proto [hunk 2]
```diff
@@ -28,4 +30,15 @@ message MsgCreateVestingAccount {
...
+
+
+// MsgCreateVestingAccount defines a message that enables creating a vesting
+// account.
+message MsgCreatePeriodicVestingAccount {
+  option (gogoproto.equal) = false;
+
+  string   from_address                    = 1 [(gogoproto.moretags) = "yaml:\"from_address\""];
+  string   to_address                      = 2 [(gogoproto.moretags) = "yaml:\"to_address\""];
+  repeated Period vesting_periods = 3 [(gogoproto.nullable) = false];
```
Comment (JimLarson) [inline proto/cosmos/vesting/v1beta1/tx.proto:43]: Specify (here or elsewhere) the semantics of the periods: is a period automatically repeating?  If I want to vest 25 quatloos per year over 4 years, do I need one Period field here, or 4?

File: x/auth/vesting/client/cli/tx.go [hunk 3]
```diff
@@ -86,3 +89,81 @@ timestamp.`,
...
+	Time  int64  `json:"unix_time"`
+}
+
+// NewMsgCreateVestingAccountCmd returns a CLI command handler for creating a
+// MsgCreateVestingAccount transaction.
+func NewMsgCreatePeriodicVestingAccountCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "create-periodic-vesting-account [to_address] [periods_json_file]",
+		Short: "Create a new vesting account funded with an allocation of tokens.",
+		Long: `Create a new vesting account funded with an allocation of tokens. This takes a destingation address and
```
Comment (JimLarson) [inline x/auth/vesting/client/cli/tx.go]: Change `destingation` to `destination`.

File: x/auth/vesting/handler_test.go [hunk 4]
```diff
@@ -91,6 +91,52 @@ func (suite *HandlerTestSuite) TestMsgCreateVestingAccount() {
...
+	testCases := []struct {
+		name      string
+		msg       *types.MsgCreatePeriodicVestingAccount
+		expectErr bool
+	}{
+		{
+			name:      "continuous vesting account already exists",
+			msg:       types.NewMsgCreatePeriodicVestingAccount(addr1, addr3, period),
+			expectErr: true,
+		},
```
Comment (JimLarson) [inline x/auth/vesting/handler_test.go:116]: Consider adding TODO comments for additional tests, invalid address, empty periods file, etc.

File: x/auth/vesting/client/cli/tx.go [hunk 5]
```diff
@@ -86,3 +89,81 @@ timestamp.`,
...
+
+			var periods []types.Period
+
+			for _, p := range inputPeriods {
+
+				amount, err := sdk.ParseCoinsNormalized(p.Coins)
+				if err != nil {
+					return err
+				}
+				period := types.Period{Length: p.Time, Amount: amount}
```
Comment (JimLarson) [inline x/auth/vesting/client/cli/tx.go]: There seems to be a mismatch between `p.Time` which is, as far as I can tell, a UNIX epoch timestamp, and the `Length` field, which `vesting.proto` implies is a duration, and is used in the context of a `start_time` field. It sure looks like this code puts a timestamp value into a duration field. If that's not the case, then either some variables or fields should be renamed, or a big comment added.

**Review by JimLarson — CHANGES_REQUESTED** [review]

File: proto/cosmos/vesting/v1beta1/tx.proto (hunk 2 above)
Comment (zmanian) [reply proto/cosmos/vesting/v1beta1/tx.proto:43]: You need 4 vesting periods of 31536000 seconds.

Comment (zmanian) [reply proto/cosmos/vesting/v1beta1/tx.proto:43]: Periods are incremental and sequential.

**Comment by zmanian** [comment]
Pushed changes to address comments.

File: x/auth/vesting/client/cli/tx.go [hunk 6]
```diff
@@ -86,3 +89,81 @@ timestamp.`,
...
+// MsgCreateVestingAccount transaction.
+func NewMsgCreatePeriodicVestingAccountCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "create-periodic-vesting-account [to_address] [periods_json_file]",
+		Short: "Create a new vesting account funded with an allocation of tokens.",
+		Long: `Create a new vesting account funded with an allocation of tokens. This takes a destination address and 
+		a period json file. 
+		Where periods.json contains:
+
+		An array of coin strings and unix epoch times for coins to vest
```
Comment (JimLarson) [inline x/auth/vesting/client/cli/tx.go:108]: Change description to "A sequence of coins and period length in seconds. Periods are sequential, in that the duration of of a period only starts at the end of the previous period. The duration of the first period starts upon account creation. For instance, the following periods.json file shows 40 "test" coins vesting annually over 4 years, where the 3rd year is a leap year." ... followed by the appropriate example.

I could imagine periods specified by the ending timestamp, or by duration from the creation of the account, so the particular choice needs to be stated more clearly, especially in external interfaces.

File: x/auth/vesting/client/cli/tx.go [hunk 7]
```diff
@@ -86,3 +89,81 @@ timestamp.`,
...
+
+[
+{
+  "coins": "10test",
+  "length_seconds":1620016673
```
Comment (JimLarson) [inline x/auth/vesting/client/cli/tx.go]: Change to a more reasonable vesting length, here and below.

File: proto/cosmos/vesting/v1beta1/tx.proto (hunk 2 above)
Comment (JimLarson) [reply proto/cosmos/vesting/v1beta1/tx.proto:43]: Okay, that checks out in the code. There's at least two other obvious ways to implement periods that use the same types (timestamp and duration-from-account-creation), so this particular choice needs to be clearly stated through all the interfaces.

Note that I don't see any prohibition of zero or negative duration, either in comments or validation checks.

File: x/auth/vesting/client/cli/tx.go [hunk 8]
```diff
@@ -86,3 +89,81 @@ timestamp.`,
...
+
+			var periods []types.Period
+
+			for _, p := range inputPeriods {
+
+				amount, err := sdk.ParseCoinsNormalized(p.Coins)
+				if err != nil {
+					return err
+				}
+				period := types.Period{Length: p.Length, Amount: amount}
```
Comment (JimLarson) [inline x/auth/vesting/client/cli/tx.go:156]: Error on zero or negative length?

File: x/auth/vesting/handler_test.go (hunk 4 above)
Comment (JimLarson) [reply x/auth/vesting/handler_test.go:116]: Also for zero or negative period length.

File: x/auth/vesting/types/msgs.go [hunk 9]
```diff
@@ -75,3 +80,55 @@ func (msg MsgCreateVestingAccount) GetSigners() []sdk.AccAddress {
...
+}
+
+// GetSignBytes returns the bytes all expected signers must sign over for a
+// MsgCreateVestingAccount.
+func (msg MsgCreatePeriodicVestingAccount) GetSignBytes() []byte {
+	return sdk.MustSortJSON(amino.MustMarshalJSON(&msg))
+}
+
+// ValidateBasic Implements Msg.
+func (msg MsgCreatePeriodicVestingAccount) ValidateBasic() error {
```
Comment (JimLarson) [inline x/auth/vesting/types/msgs.go:118]: This might be a good place to check for zero or negative period lengths.

**Review by JimLarson — CHANGES_REQUESTED** [review]
Thanks for the explanation! Since there are other reasonable ways to implement periods that all use int in the spec, I've suggested some good places for comments. It would also be good to raise an error on zero or negative period length.

File: proto/cosmos/vesting/v1beta1/tx.proto (hunk 2 above)
Comment (zmanian) [reply proto/cosmos/vesting/v1beta1/tx.proto:43]: Add length checks to in ValidateBasic

**Comment by zmanian** [comment]
Resolved the major outstanding issues.

**Review by JimLarson — APPROVED** [review]
